usr/src/uts/common/fs/zfs/arc.c

```diff
@@ -3069,97 +3074,113 @@ arc_evict_state(arc_state_t *state, uint64_t spa, int64_t bytes,
 	num_sublists = multilist_get_num_sublists(ml);
 
 	/*
-	 * If we've tried to evict from each sublist, made some
-	 * progress, but still have not hit the target number of bytes
-	 * to evict, we want to keep trying. The markers allow us to
-	 * pick up where we left off for each individual sublist, rather
-	 * than starting from the tail each time.
+	 * Do the job for each type of headers
+	 * A header type of 0 is used to indicate that we first only
+	 * evict headers which have L2 headers.
+	 * A header type of 1 does the same but tries to find more
+	 * such headers ignoring zfs_arc_evict_batch_limit.
+	 * A header type of 2 catches both L1 & L2 headers.
 	 */
-	markers = kmem_zalloc(sizeof (*markers) * num_sublists, KM_SLEEP);
-	for (int i = 0; i < num_sublists; i++) {
-		markers[i] = kmem_cache_alloc(hdr_full_cache, KM_SLEEP);
+	for (int h = 0; h < 3 && (total_evicted < bytes || bytes == ARC_EVICT_ALL); h++) {
```

Comment:
I think this would be more clear if we used an enum for `h`, instead of magic numbers.